gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java

```diff
@@ -95,7 +108,16 @@ public void run()
         .createDefaultTopLevelBroker(jobConfig, GobblinScopeTypes.GLOBAL.defaultScopeInstance())) {
       SharedResourcesBroker<GobblinScopeTypes> jobBroker = getJobBroker(_jobState, globalBroker);
 
-      _taskAttempt = _taskAttemptBuilder.build(workUnits.iterator(), _jobId, _jobState, jobBroker);
+      // Secure atomicity of taskAttempt's execution.
+      // Signaling blocking threads if any whenever taskAttempt is nonNull.
+      _lock.lock();
+      try {
+        _taskAttempt = _taskAttemptBuilder.build(getWorkUnits().iterator(), _jobId, _jobState, jobBroker);
+        _taskAttemptBuilt.signal();
+      } finally {
+        _lock.unlock();
+      }
+
       _taskAttempt.runAndOptionallyCommitTaskAttempt(GobblinMultiTaskAttempt.CommitPolicy.IMMEDIATE);
```

Comment:
why do we move this out of the lock? What will happen if _taskAttemp is initiated but not run, and cancel happen and then call _taskAttempt.runAndOptionallyCommitTaskAttempt()

Reply:
I think it is fine, the way that cancel works is, it call `shutdownTasks()` which actually sets a flag for each task, indicating the task has been terminated externally. This signal will be checked in extractor as it pull data from upstream, whenever it is being detected, it accomplish the ongoing loop of pulling data.  So the `run()` method within `runAndOptionallyCommitTaskAttempt` should still return. 

But I don't have test case to verify this though, I might be too aggressive for pruning critical section since it is not performance sensitive at all. Will put it back.